### PR TITLE
AP_OSD: added OSD_TYPE2 param to enable a second OSD backend

### PR DESCRIPTION
**Pull request by yaapu**
This will enable a second OSD backend in parallel with the default one.
A check is in place to only allow certain backend combinations, for instance two MSP serial backends are not supported as well as two MAX or SITL backends.

stats and panels are shared between the two instances.

### CODE REVIEW

**Comment by tridge** [comment]
I like the idea behind this, it just needs more testing to especially ensure that normal single OSD works fine, and at least one user confirming that 2 OSD works

File: libraries/AP_OSD/AP_OSD.cpp [hunk 1]
```diff
@@ -223,6 +223,14 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
     AP_SUBGROUPINFO2(screen[2], "3_", 29, AP_OSD, AP_OSD_Screen),
     AP_SUBGROUPINFO2(screen[3], "4_", 30, AP_OSD, AP_OSD_Screen),
 
+    // @Param: 2_TYPE
```
Comment (Hwurzburg) [inline libraries/AP_OSD/AP_OSD.cpp:226]: should be named OSD_TYPE2....to avoid grouping in the second screen parameter list

**Review by Hwurzburg — COMMENTED** [review]
I had a plane with hardware for internal OSD + DJI Air unit...
I have screen 1 optimized for DJI
and second screen optimized for Internal OSD
I had been running this autopilot with OSD_TYPE=3 and the DJI googles

I loaded your PR and only set OSD_TYPE=1 and changed to screen two for the internal OSD (both screens have 900-2000PWM and I disable one and enable the other to aoivd having to have a TX on to switch screens)
cycling power.....it is now still outputing DJI and is still on the first screen! strange!!
at this point I stopped testing....must have something wrong

File: libraries/AP_OSD/AP_OSD.cpp (hunk 1 above)
Comment (yaapu) [reply libraries/AP_OSD/AP_OSD.cpp:226]: ok param name changed

**Comment by yaapu** [comment]
> I had a plane with hardware for internal OSD + DJI Air unit... I have screen 1 optimized for DJI and second screen optimized for Internal OSD I had been running this autopilot with OSD_TYPE=3 and the DJI googles
> 
> I loaded your PR and only set OSD_TYPE=1 and changed to screen two for the internal OSD (both screens have 900-2000PWM and I disable one and enable the other to aoivd having to have a TX on to switch screens) cycling power.....it is now still outputing DJI and is still on the first screen! strange!! at this point I stopped testing....must have something wrong

mhmm, weird, I have it working on an amnibusf4 v3, main OSD is analog so OSD_TYPE=1  and second is MSP DisplayPort with OSD_TYPE2=5 and SERIAL1_PROTOCOL=42

**Comment by Hwurzburg** [comment]
I have repeated on two different vehicles....from fresh code install...I have two screens with full range on PWM...one optimized for DJI one for internal....enable only one screen at a time...
OSD_TYPE=3
OSD_TYPE2=0     on Screen1 optimized for DJI goggles....works fine
then
OSD_TYPE2=1  power cycle and the internal OSD works but screen is messed up because its on Screen 1
then reverse types putting 1 on first and 3 on second, switch to screen 2 optimized for internal OSD, power cycle.....internal OSD no longer works....DJI googles work and are on SCREEN1 which is DISABLED!

Switch OSD_TYPE=3 and OSD_TYPE2=0...switch to SCREEN2..power cycle....no change to screen, stays on SCREEN1!

Switch OSD_TYPE=1 and OSD_TYPE2=0....power cycle.....DJI OSD still enabled and on screen 1 even though its disabled in params and on screen2!!

**Comment by Hwurzburg** [comment]
Update...so testing with Plane 4.1.4...there seems to be a regression bug that DJI MSP will not screen switch with TX.....
also to switch screens you must use the TX, you cannot just set screen2 enabled and screen1 disabled and get screen2...this should be fixed also and is a Stable issue...so that clears up the screen mystery....
AND once the DJI  backend is enabled, it remains enabled even if the internal OSD is active! so dual OSD was already in Plane 4.1.4! by mistake....


Testing with 4.2 version a month old...the screen switching works, but DJI OSD is still persistent, and switches screens along with the internal OSD...

**Comment by yaapu** [comment]
> but DJI OSD is still persistent,
this is not a real bug but rather a side effect :-)
the DJI backend is updated by the MSP thread not by the OSD thread, the OSD thread is there only to update the stats but otherwise does nothing, so we always had this and OSD_TYPE=3 actually activates a light OSD backend that only does the stats, you can have MSP DJI with OSD_TYPE=0, but you would not get the stats.
So we kind of had multiple OSDs but only with MSP as second OSD, the combination Analog + DisplayPort was not supported

**Comment by Hwurzburg** [comment]
Tested the following combinations:
image
so one combo does not work, but its inverse does

**Comment by Hwurzburg** [comment]
I was looking over your backend exclusions and realized that the case of 5/3 was also an excluded 2nd backend but MSP/DJI OSD worked because Displayport WAS active so MSP could use the screen setup params....probably having two msp like backends like DISPLAYPORT and MSP is not required and the current exclusion is correct....however, the GCS message happens so early that it is lost to the user....I think a prearm misconfiguration message might be better.... so both 3/5 and 5/3 would give a prearm reconfig message and we can explain in the wiki

I think moving the gcs to a prearm is all that this needs. along with a wiki section explaining acceptable combinations....

**Comment by IamPete1** [comment]
I agree, the combinations seem a little odd, but if we can have a nice pre-arm that give a good error its OK.  

A future enhancement might be allow them each to display subset of the screens.

**Comment by Hwurzburg** [comment]
@yaapu I added a prearm ...I did not squash so its easier to review the changes step by step....works in SITL...will test on boards next...

**Comment by Hwurzburg** [comment]
okay have tested on the bench and everything is working now for the prearm...removed the gsc message on osd init as its invisible to most users ...even the weird case of TYPE=0 and TYPE2 is assigned but invisible (its not instantiated, must have osd1 in order to show osd2 as an expansion)

File: libraries/AP_OSD/AP_OSD.cpp [hunk 2]
```diff
@@ -263,7 +272,26 @@ AP_OSD::AP_OSD()
 
 void AP_OSD::init()
 {
-    switch (osd_types(osd_type.get())) {
+    const AP_OSD::osd_types types[OSD_MAX_INSTANCES] = {osd_types(osd_type.get()), osd_types(osd_type2.get())};
+    for (uint8_t instance = 0; instance < OSD_MAX_INSTANCES; instance++) {
+        if (init_backend(types[instance], instance)) {
+            if (instance == 0) {
+                hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_OSD::osd_thread, void), "OSD", 1280, AP_HAL::Scheduler::PRIORITY_IO, 1);
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD.cpp:279]: this means we don't create thread if 1st instance fails to init, add thread of _backend_count is > 0

File: libraries/AP_Arming/AP_Arming.h [hunk 3]
```diff
@@ -38,6 +38,7 @@ class AP_Arming {
         ARMING_CHECK_AUX_AUTH    = (1U << 17),
         ARMING_CHECK_VISION      = (1U << 18),
         ARMING_CHECK_FFT         = (1U << 19),
+        ARMING_CHECK_OSD         = (1U << 11),
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.h:41]:
```suggestion
        ARMING_CHECK_OSD         = (1U << 20),
```

remember we only have ~28 of these, however :-)

File: libraries/AP_OSD/AP_OSD.cpp [hunk 4]
```diff
@@ -232,6 +232,15 @@ const AP_Param::GroupInfo AP_OSD::var_info[] = {
...
     AP_SUBGROUPINFO2(screen[3], "4_", 30, AP_OSD, AP_OSD_Screen),
 #endif
+
+    // @Param: _TYPE2
+    // @DisplayName: OSD type 2
+    // @Description: OSD type 2. TXONLY makes the OSD parameter selection available to other modules even if there is no native OSD support on the board, for instance CRSF.
+    // @Values: 0:None,1:MAX7456,2:SITL,3:MSP,4:TXONLY,5:MSP_DISPLAYPORT
+    // @User: Standard
+    // @RebootRequired: True
+    AP_GROUPINFO("_TYPE2", 32, AP_OSD, osd_type2, 0),
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD.cpp:242]: Has this been tested in SITL to make sure the ID is OK?

File: libraries/AP_OSD/AP_OSD.cpp [hunk 5]
```diff
@@ -263,7 +272,26 @@ AP_OSD::AP_OSD()
 
 void AP_OSD::init()
 {
-    switch (osd_types(osd_type.get())) {
+    const AP_OSD::osd_types types[OSD_MAX_INSTANCES] = {osd_types(osd_type.get()), osd_types(osd_type2.get())};
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD.cpp:275]:
```suggestion
    const AP_OSD::osd_types types[OSD_MAX_INSTANCES] = {
        osd_types(osd_type.get()),
        osd_types(osd_type2.get()),
    };
```

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_OSD/AP_OSD_MAX7456.h [hunk 6]
```diff
@@ -39,7 +39,21 @@ class AP_OSD_MAX7456 : public AP_OSD_Backend
...
     // return a correction factor used to display angles correctly
-   float get_aspect_ratio_correction() const override;
+    float get_aspect_ratio_correction() const override;
+
+    bool is_compatible_with_backend_type(AP_OSD::osd_types type) const override {
+        switch(type) {
+        case AP_OSD::osd_types::OSD_MAX7456:
+        case AP_OSD::osd_types::OSD_SITL:
+            return false;
+        default:
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_MAX7456.h:49]: Don't use a `default:` here; enumerate the other types.  That forces someone creating a new backend type to actually specify the compatability here

Similarly elsewhere.

File: libraries/AP_OSD/AP_OSD_MAX7456.h [hunk 7]
```diff
@@ -39,7 +39,21 @@ class AP_OSD_MAX7456 : public AP_OSD_Backend
     void clear() override;
 
     // return a correction factor used to display angles correctly
-   float get_aspect_ratio_correction() const override;
+    float get_aspect_ratio_correction() const override;
```
Comment (peterbarker) [inline libraries/AP_OSD/AP_OSD_MAX7456.h:42]:
```suggestion
   float get_aspect_ratio_correction() const override;
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Partial review only

File: libraries/AP_OSD/AP_OSD.cpp (hunk 2 above)
Comment (yaapu) [reply libraries/AP_OSD/AP_OSD.cpp:279]: auch, you're right

File: libraries/AP_Arming/AP_Arming.h (hunk 3 above)
Comment (Hwurzburg) [reply libraries/AP_Arming/AP_Arming.h:41]: sorry for the typo

File: libraries/AP_OSD/AP_OSD_MAX7456.h (hunk 6 above)
Comment (yaapu) [reply libraries/AP_OSD/AP_OSD_MAX7456.h:49]: ok, I had to add a "return false" at the end for the compiler would complain about unreachable code but this way if the enum is expanded the compiler will complain as well so good to go.

File: libraries/AP_OSD/AP_OSD.cpp (hunk 4 above)
Comment (Hwurzburg) [reply libraries/AP_OSD/AP_OSD.cpp:242]: if you mean has the SITL osd been tested....yes....it does work as 2nd type in SITL for its compatible cases

File: libraries/AP_OSD/AP_OSD_MAX7456.h (hunk 7 above)
Comment (yaapu) [reply libraries/AP_OSD/AP_OSD_MAX7456.h:42]: ok

File: libraries/AP_OSD/AP_OSD.cpp (hunk 4 above)
Comment (yaapu) [reply libraries/AP_OSD/AP_OSD.cpp:242]: Yes, all combinations but @Hwurzburg offered to retest

**Comment by Hwurzburg** [comment]
retested

**Comment by Hwurzburg** [comment]
@yaapu I went ahead and squashed it so it can CI by Devcall

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
happy for this to be merged when @peterbarker approved

**Comment by tridge** [comment]
nice feature!

**Comment by Hwurzburg** [comment]
tested the new arming check box for OSD....needed a small change...tested now...

**Pull request merged**